src/ff1.rs

```diff
@@ -90,6 +98,54 @@ impl RadixOps<u16> for u16 {
     }
 }
 
+/// A radix 2^i for i in [1..16). It does not use floating-point arithmetic.
+#[derive(Debug, PartialEq)]
+pub struct PowerTwoRadix {
+    radix: u16,
+    log_radix: u8,
+}
+
+impl PowerTwoRadix {
+    pub fn from(radix: u16) -> Result<Self, ()> {
+        let mut tmp = radix;
+        let mut log_radix = 0;
+        let mut found = false;
+        for i in 0..16 {
+            if tmp & 1 != 0 {
+                // Only a single bit should be set
+                if found {
+                    return Err(());
+                }
+                log_radix = i;
+                found = true;
+            }
+            tmp >>= 1;
+        }
+        match log_radix {
+            0 => Err(()),
+            _ => Ok(PowerTwoRadix { radix, log_radix }),
+        }
+    }
+}
+
+impl RadixOps for PowerTwoRadix {
+    fn check_in_range(&self, n: u32) -> bool {
+        n >> self.log_radix == 0
```

Comment:
``n < self.radix`` is simpler.

Reply:
Whoops, I completely missed that! I'll address it in a follow-up PR.